tpi_rs/src/middleware/persistency/app_persistency.rs

```diff
@@ -0,0 +1,282 @@
+}
+
+impl PersistencyBuilder {
+    /// Add a key to the key/value store. Attempting to access keys that are not
+    /// registered with this function will result in an error.
+    pub fn register_key<T>(mut self, key: &'static str, default: T) -> Self
+    where
+        T: Send + serde::Serialize,
+    {
+        self.keys.push((key, bincode::serialize(&default).unwrap()));
+        self
+    }
+
+    /// The [ApplicationPersistency] contains a write mechanism that writes the
+    /// key/value store back to the file-system. This happens when on a timeout
+    /// occurrence started from the last write. This function disables the write
+    /// on timeout. The key/value store only gets written when the
+    /// [ApplicationPersistency] is dropped.
+    pub fn disable_write_on_timeout(mut self) -> Self {
+        self.write_timeout = None;
+        self
+    }
+
+    /// Construct an [ApplicationPersistency] object.
+    pub async fn build(self) -> anyhow::Result<ApplicationPersistency> {
+        ApplicationPersistency::new(self.keys, BIN_DATA, self.write_timeout).await
+    }
+}
+
+impl Default for PersistencyBuilder {
+    fn default() -> Self {
+        Self {
+            keys: Vec::new(),
+            write_timeout: Some(WRITE_BACK_TIMEOUT),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct MonitorContext {
+    pub file: PathBuf,
+    pub inner: PersistencyStore,
+}
+
+impl MonitorContext {
+    pub async fn commit_to_file(&self) -> anyhow::Result<MonitorEvent> {
+        let mut new = self.file.clone();
+        new.set_extension("new");
+
+        let pending = OpenOptions::new()
+            .create(true)
+            .truncate(true)
+            .write(true)
+            .open(&new)
+            .await?;
+        self.inner.write(pending.into_std().await).await?;
+
+        tokio::fs::rename(&new, &self.file)
+            .await
+            .context("permanent damaged persistency binary")?;
+
+        Ok(MonitorEvent::PersistencyWritten)
+    }
+
+    pub async fn sync_all(&self) -> anyhow::Result<()> {
+        self.commit_to_file().await?;
+
+        let file = File::open(&self.file).await?;
+        file.sync_all().await?;
+        Ok(())
+    }
+}
+
+/// This struct represents a concrete key/value store used by the bmcd. It sets
+/// up a key/value store on the given path. It monitors the store for changes
+/// and writes back all changes when no update to the store was detected for a
+/// given amount of time.(`write_timeout`) when `None` is passed as
+/// `write_timeout` the key/value store only gets written when the
+/// [ApplicationPersistency] is dropped.
+#[derive(Debug)]
+pub struct ApplicationPersistency {
+    context: Arc<MonitorContext>,
+}
+
+impl ApplicationPersistency {
+    pub async fn new<I, P>(
+        keys_with_default: I,
+        path: P,
+        write_timeout: Option<Duration>,
+    ) -> anyhow::Result<Self>
+    where
+        I: IntoIterator<Item = (&'static str, Vec<u8>)>,
+        P: Into<PathBuf>,
```

Comment:
These two types should also be `Send` to be safely sendable between await points

```diff
-        I: IntoIterator<Item = (&'static str, Vec<u8>)>,
-        P: Into<PathBuf>,
+        I: IntoIterator<Item = (&'static str, Vec<u8>)> + Send,
+        P: Into<PathBuf> + Send,
```

Same for binary_persistency.rs:112

Reply:
`new` is not a `async` function. Therefore the arguments do not travel over await points, `Send` does not apply here.

Reply:
> Same for binary_persistency.rs:112

Calling `async` functions does not imply that such functions can be executed across threads. `async` functions are always executed in the current thread's context and can only be pushed to other threads by explicit operations. e.g. `tokio::spawn`.

You can see `async` functions as single-threaded state machines that yield back control to the event loop of the thread such function runs on. e.g., they express a way to execute a blocking function non-blocking.

This means that `Send` is not required here.